File: yukicoder/submit.py
#!/usr/bin/env python3
"""Submit a solution file to yukicoder.
"""

import sys
import json
import os
import cerberus
import requests
import yaml

def read_yukicoder_config(yukicoder_config_path):
    """Read yukicoder-specific config
    """
    with open(yukicoder_config_path, encoding='utf-8') as file:
        config = yaml.safe_load(file)
    val = cerberus.Validator({'api_key': {'type': 'string'}})
    if not val.validate(config):
        print('Invalid config: {} {}'.format(yukicoder_config_path, val.errors),
              file=sys.stderr)
        sys.exit(1)
    return config


def read_language_config(language_config_path):
    """Read language config
    """
    with open(language_config_path, encoding='utf-8') as file:
        languages = yaml.safe_load(file)
    # Cerberus does not support top-level list values;
    # therefore, we use a hack here: wrapping languages with a dummy key-value pair.
    value = {'dummy': languages}
    item_schema = {
        'extension': {'type': 'string'},
        'yukicoder_name': {'type': 'string'}
    }
    value_schema = {
        'dummy': {
            'type': 'list',
            'schema': {
                'type': 'dict',
                'schema': item_schema,
                'allow_unknown': True
            }
        }
    }
    val = cerberus.Validator(value_schema)
    if not val.validate(value):
        print('Invalid languages config: {} {}'.format(language_config_path, val.errors),
              file=sys.stderr)
        sys.exit(1)
    return languages


def main():
    """main
    """
    if len(sys.argv) != 2:
        print("./submit.py PROBLEM_ID.ext")
        sys.exit(1)

    filename = sys.argv[1]

    script_dir = os.path.dirname(os.path.abspath(__file__))
    yukicoder_config_path = os.path.join(script_dir, 'yukicoder_config')
    language_config_path = os.path.join(script_dir, '..', 'languages.yml')

    yukicoder_config = read_yukicoder_config(yukicoder_config_path)

    languages = read_language_config(language_config_path)


    (problem_name, extension) = os.path.splitext(filename)
    if extension == '':
        print("\x1b[34minvalid extension\x1b[0m")
        sys.exit(1)

    lang = next(x for x in languages if x['extension'] == extension[1:])['yukicoder_name']
    with open(filename, encoding='utf-8') as file:
        source = file.read()

    api_key = yukicoder_config['api_key']

    url = f'https://yukicoder.me/api/v1/problems/no/{problem_name}'
    resp = requests.get(url, headers={
        'Authorization': "bearer " + api_key,
        'Accept': 'application/json'
    })
    response = json.loads(resp.text)

    # Find problem_id.
    if 'ProblemId' not in response:
        print('Invalid response: {}', response)
        sys.exit(2)
    problem_id = response['ProblemId']

    post_url = f'https://yukicoder.me/api/v1/problems/{problem_id}/submit'

    print("Submitting \x1b[34m{}\x1b[0m as \x1b[34m{}\x1b[0m".format(filename, lang))
    print("\tto \x1b[32m{}\x1b[0m (problem_id = {})".format(problem_name, problem_id))

    resp = requests.post(post_url, {
        'lang': lang, 'source': source,
    }, headers={
        'Authorization': "bearer " + api_key,
        'Accept': 'application/json'
    })

    response = json.loads(resp.text)

    print(response)

    if 'SubmissionId' not in response:
        print(response)
        print("\x1b[34mSubmission unsuccessful\x1b[0m")
        sys.exit(1)

main()

File: yukicoder/submit.go
package main

// Compile: go build -o submit submit.go
// Run: ./submit PROBLEM_ID.ext

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"mime/multipart"
	"net/http"
	"os"
	"path"
	"strings"

	"gopkg.in/yaml.v2"
)

type yukicoderConfig struct {
	ApiKey string `yaml:"api_key"`
}

type languageConfigEntry struct {
	Extension     string `yaml:"extension"`
	YukicoderName string `yaml:"yukicoder_name"`
}

func readYukicoderConfig(yukicoderConfigPath string) (*yukicoderConfig, error) {
	yamlContent, err := ioutil.ReadFile(yukicoderConfigPath)
	if err != nil {
		return nil, err
	}
	var result yukicoderConfig
	if err := yaml.Unmarshal(yamlContent, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

func readLanguageConfig(languageConfigPath string) ([]languageConfigEntry, error) {
	yamlContent, err := ioutil.ReadFile(languageConfigPath)
	if err != nil {
		return nil, err
	}
	var result []languageConfigEntry
	if err := yaml.Unmarshal(yamlContent, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func getProblemIDFromProblemName(problemName string) (int, error) {
	url := fmt.Sprintf("https://yukicoder.me/api/v1/problems/no/%s", problemName)
	resp, err := http.Get(url)
	if err != nil {
		return 0, err
	}
	if resp.StatusCode != 200 {
		return 0, fmt.Errorf("StatusCode = %d != 200", resp.StatusCode)
	}
	var respData struct {
		ProblemID int `json:"ProblemId"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&respData); err != nil {
		return 0, err
	}
	return respData.ProblemID, nil
}

// postCode returns SubmissionId
func postCode(problemID int, language string, code []byte, apiKey string) (uint64, error) {
	url := fmt.Sprintf("https://yukicoder.me/api/v1/problems/%d/submit", problemID)

	client := &http.Client{}

	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	if err := writer.WriteField("lang", language); err != nil {
		log.Println("Populating lang failed")
		return 0, err
	}
	if err := writer.WriteField("source", string(code)); err != nil {
		log.Println("Populating source failed")
		return 0, err
	}
	writer.Close()

	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		log.Println("NewRequest failed")
		return 0, err
	}
	req.Header.Add("Authorization", "bearer "+apiKey)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", writer.FormDataContentType())

	resp, err := client.Do(req)
	if err != nil {
		log.Println("Request failed")
		return 0, err
	}

	var respData struct {
		SubmissionID uint64 `json:"SubmissionId"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&respData); err != nil {
		log.Println("Parsing JSON failed")
		return 0, err
	}
	return respData.SubmissionID, nil
}

func main() {
	if len(os.Args) != 2 {
		log.Fatalf("Usage: ./submit PROBLEM_ID.ext\n")
	}

	log.Print("Start submitting...")
	filename := os.Args[1]

	exec, err := os.Executable()
	if err != nil {
		log.Fatal(err)
	}
	scriptDir := path.Dir(exec)
	yukicoderConfigPath := path.Join(scriptDir, "yukicoder_config")
	languageConfigPath := path.Join(scriptDir, "..", "languages.yml")
	yukicoderConfig, err := readYukicoderConfig(yukicoderConfigPath)
	if err != nil {
		log.Fatal(err)
	}
	languages, err := readLanguageConfig(languageConfigPath)
	if err != nil {
		log.Fatal(err)
	}

	extension := path.Ext(filename)
	if extension == "" {
		log.Fatalln("\x1b[34minvalid extension\x1b[0m")
	}
	problemName := strings.TrimSuffix(filename, extension)
	var language string
	for _, entry := range languages {
		if "."+entry.Extension == extension {
			language = entry.YukicoderName
		}
	}
	if language == "" {
		log.Fatalf("Language for extension %s not found in languages.yml", extension)
	}
	code, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	problemID, err := getProblemIDFromProblemName(problemName)
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("Submitting \x1b[34m%s\x1b[0m as \x1b[34m%s\x1b[0m", filename, language)
	log.Printf("\tto \x1b[32m%s\x1b[0m (problem_id = %d)", problemName, problemID)
	submissionID, err := postCode(problemID, language, code, yukicoderConfig.ApiKey)
	if err != nil {
		log.Fatal("Submission unsuccessful: ", err)
	}
	log.Printf("Done. SubmissionId = %d", submissionID)
}
